### PR TITLE
feat: add hyperparameter tuning v2

### PR DESCRIPTION
**Pull request by MartinBernstorff**
# Notes for reviewers
Reviewers can skip X, but should pay attention to Y.

### CODE REVIEW

**Comment by MartinBernstorff** [comment]
Fixes #421.

File: psycop/common/model_training_v2/hyperparameter_suggester/hyperparameter_suggester.py [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+    Args:
+        base_cfg: The base config, a tree of options.
+        trial: The optuna trial to use for suggesting hyperparameters.
+    """
+    cfg = base_cfg.copy()
+    process_all_values_in_dict(d=cfg, trial=trial)
+    return cfg
+
+
+def process_all_values_in_dict(d: dict[str, Any], trial: optuna.Trial):
```
Comment (HLasse) [inline psycop/common/model_training_v2/hyperparameter_suggester/hyperparameter_suggester.py:45]: Rename to something more descriptive, e.g. `suggest_hparams_in_tree` (not a great example..) -- too late in the day for great creativity.. 🧠

File: psycop/common/model_training_v2/hyperparameter_suggester/hyperparameter_suggester.py [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+    for key, value in d.items():
+        match value:
+            case dict():
+                process_all_values_in_dict(value, trial=trial)
+            case SearchSpace():
+                d[key] = value.suggest_hyperparameters(trial=trial)
+            case _:
```
Comment (HLasse) [inline psycop/common/model_training_v2/hyperparameter_suggester/hyperparameter_suggester.py:52]: are we assuming all keys return either a dict or a SearchSpace here? What happens to params that should not be searched over?

File: psycop/common/model_training_v2/hyperparameter_suggester/hyperparameter_suggester.py [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+
+
+    Base-cfg is a config dict representing a tree. For all nodes in the tree that are of type SearchSpace, replace the node with a suggested set of a hyperparameters.
+
+    Args:
+        base_cfg: The base config, a tree of options.
+        trial: The optuna trial to use for suggesting hyperparameters.
+    """
+    cfg = base_cfg.copy()
+    process_all_values_in_dict(d=cfg, trial=trial)
```
Comment (HLasse) [inline psycop/common/model_training_v2/hyperparameter_suggester/hyperparameter_suggester.py:41]: hm, so we're modifying the cfg dict in-place. Feels a bit iffy, but the recursion is neat.. Would prefer a solution without in-place modification as it's much easier to comprehend

File: psycop/common/model_training_v2/hyperparameter_suggester/suggesters/logistic_regression_suggester.py [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+            high=self.high,
+            log=self.logarithmic,
+        )
+
+
+class LogisticRegressionSuggester(Suggester):
+    # XXX: Should we move this to the logistic regression creator, and then have the creator also subclass suggester?
+    # Perhaps that makes it hard to decorate, though?
+
+    # XXX: When using confection, this class becomes _much_ easier to initialise if initialised with tuples of (low, high, logarithmic). Then we can recast to FloatSpace. Otherwise, we will have to initialise a dataclass every time. Do we agree here?
```
Comment (HLasse) [inline psycop/common/model_training_v2/hyperparameter_suggester/suggesters/logistic_regression_suggester.py:32]: 👍

File: psycop/common/model_training_v2/hyperparameter_suggester/suggesters/logistic_regression_suggester.py [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+        return trial.suggest_float(
+            name=name,
+            low=self.low,
+            high=self.high,
+            log=self.logarithmic,
+        )
+
+
+class LogisticRegressionSuggester(Suggester):
+    # XXX: Should we move this to the logistic regression creator, and then have the creator also subclass suggester?
```
Comment (HLasse) [inline psycop/common/model_training_v2/hyperparameter_suggester/suggesters/logistic_regression_suggester.py:29]: I agree that it's appealing to have them together, but I think you're right that decorating them will be hard then

File: psycop/common/model_training_v2/hyperparameter_suggester/suggesters/test_suggesters.py [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+def suggester_tester(suggester: Suggester) -> TestSuggestion:
+    """Test utility function which ensures that the suggester:
+    1. Interfaces correctly with Optuna
+    2. Can be resolved from the BaselineRegistry
+    """
+    sampler = optuna.samplers.RandomSampler()
+
+    with StorageSupplier("inmemory") as storage:
+        study = optuna.create_study(storage=storage, sampler=sampler)
+        trial = optuna.Trial(study, study._storage.create_new_trial(study._study_id))  # type: ignore
```
Comment (HLasse) [inline psycop/common/model_training_v2/hyperparameter_suggester/suggesters/test_suggesters.py:38]: Would be great to this as a fixture or something easier to use for other tests, but can be delegated to different PR

Comment (HLasse) [reply psycop/common/model_training_v2/hyperparameter_suggester/suggesters/test_suggesters.py:38]: Neat testing function though!

File: psycop/common/model_training_v2/hyperparameter_suggester/test_hyperparameter_suggester.py [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+)
+from psycop.common.model_training_v2.hyperparameter_suggester.suggesters.logistic_regression_suggester import (
+    LogisticRegressionSuggester,
+)
+from psycop.common.model_training_v2.hyperparameter_suggester.suggesters.test_suggesters import (
+    float_space_for_test,
+)
+
+
+class TestHyperparameterSuggster:
```
Comment (HLasse) [inline psycop/common/model_training_v2/hyperparameter_suggester/test_hyperparameter_suggester.py:18]:
```suggestion
class TestHyperparameterSuggester:
```

File: psycop/common/model_training_v2/hyperparameter_suggester/hyperparameter_suggester.py [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+
+        # XXX: Replace __repr__ with something better
+        suggester_names = list(suggester_dict.keys())
+        suggester_name: str = trial.suggest_categorical("suggester", suggester_names)  # type: ignore # We know this is a string, because it must suggest from the suggester_names. Optuna should type-hint with a generic, but haven't. MB has created an issue here: https://github.com/optuna/optuna/issues/5104
+
+        suggester = suggester_dict[suggester_name]
+        return suggester.suggest_hyperparameters(trial=trial)
+
+
+def hyperparameter_suggester(
```
Comment (HLasse) [inline psycop/common/model_training_v2/hyperparameter_suggester/hyperparameter_suggester.py:27]:
```suggestion
def suggest_hyperparameters_from_cfg(
```

File: psycop/common/model_training_v2/hyperparameter_suggester/test_hyperparameter_suggester.py [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+        with StorageSupplier("inmemory") as storage:
+            study = optuna.create_study(storage=storage, sampler=sampler)
+            trial = optuna.Trial(
+                study,
+                study._storage.create_new_trial(study._study_id),  # type: ignore
+            )
+
+            return hyperparameter_suggester(base_cfg=base_cfg, trial=trial)
+
+    def test_hyperparameter_suggester(self):
```
Comment (HLasse) [inline psycop/common/model_training_v2/hyperparameter_suggester/test_hyperparameter_suggester.py:31]: consider parametrizing this so we can easily test multiple different suggesters. But looks good!

**Review by HLasse — APPROVED** [review]
Bunch of minor stuff, but nice, Martin! Looking forward to testing this out!

File: psycop/common/model_training_v2/hyperparameter_suggester/hyperparameter_suggester.py (hunk 3 above)
Comment (MartinBernstorff) [reply psycop/common/model_training_v2/hyperparameter_suggester/hyperparameter_suggester.py:41]: I agree in general, but I think it's OK since we wrap it, and noone should call it directly. I'll make it private to make that clearer.

File: psycop/common/model_training_v2/hyperparameter_suggester/suggesters/test_suggesters.py (hunk 6 above)
Comment (MartinBernstorff) [reply psycop/common/model_training_v2/hyperparameter_suggester/suggesters/test_suggesters.py:38]: I'm confused about the value-add of a fixture here? Since it's a utility-function, isn't an import as easy to read? :-)

File: psycop/common/model_training_v2/hyperparameter_suggester/suggesters/logistic_regression_suggester.py [hunk 10]
```diff
@@ -25,14 +26,37 @@ def suggest(self, trial: optuna.Trial, name: str) -> float:
         )
 
 
+@BaselineRegistry.estimator_steps.register("mock_suggester")
+class MockSuggester(Suggester):
```
Comment (MartinBernstorff) [inline psycop/common/model_training_v2/hyperparameter_suggester/suggesters/logistic_regression_suggester.py:30]: @HLasse This should make tests more robust. Thought you might enjoy it 👍

File: psycop/common/model_training_v2/hyperparameter_suggester/suggesters/test_suggesters.py [hunk 11]
```diff
@@ -47,10 +48,18 @@ def suggester_tester(suggester: Suggester) -> TestSuggestion:
 def test_logistic_regression_suggester():
     result = suggester_tester(
         suggester=LogisticRegressionSuggester(
-            C=(0,1,False),
-            l1_ratio=(0,1,False),
+            C_low=0.1,
+            C_high=1,
+            C_log=False,
```
Comment (MartinBernstorff) [inline psycop/common/model_training_v2/hyperparameter_suggester/suggesters/test_suggesters.py:53]: Had to change the interface because confection cannot resolve tuples from the .cfg

File: psycop/common/model_training_v2/hyperparameter_suggester/hyperparameter_suggester.py [hunk 12]
```diff
@@ -0,0 +1,53 @@
...
+    """Suggest hyperparameters in a config.
+
+
+    Base-cfg is a config dict representing a tree. For all nodes in the tree that are of type SearchSpace, replace the node with a suggested set of a hyperparameters.
+
+    Args:
+        base_cfg: The base config, a tree of options.
+        trial: The optuna trial to use for suggesting hyperparameters.
+    """
+    cfg = base_cfg.copy()
```
Comment (HLasse) [inline psycop/common/model_training_v2/hyperparameter_suggester/hyperparameter_suggester.py:39]: deepcopy

**Review by HLasse — APPROVED** [review]
❤️
